File: debop4k-web/src/main/java/debop4k/web/TomcatLauncher.java
//package debop4k.web;
//
//import debop4k.core.Charsets;
//import lombok.extern.slf4j.Slf4j;
//import org.apache.catalina.Context;
//import org.apache.catalina.connector.Connector;
//import org.apache.catalina.core.AprLifecycleListener;
//import org.apache.catalina.core.JreMemoryLeakPreventionListener;
//import org.apache.catalina.core.ThreadLocalLeakPreventionListener;
//import org.apache.catalina.mbeans.GlobalResourcesLifecycleListener;
//import org.apache.catalina.startup.Tomcat;
//import org.springframework.boot.CommandLineRunner;
//
//import java.io.File;
//
///**
// * Tomcat Server 를 Embedded 방식으로 실행할 수 있도록 지원합니다.
// * Stand alone 으로 실행해야 할 Web Application 이나 테스트용 Web Application을 만들 때 사용합니다.
// *
// * @author sunghyouk.bae@gmail.com
// */
//@Slf4j
//public class TomcatLauncher implements CommandLineRunner {
//
//  public int getPort() { return 8080; }
//  public String getContextPath() { return "/"; }
//  public String getResourceBase() { return "src/main/webapp"; }
//  public String getProtocol() { return "org.apache.coyote.http11.Http11Protocol"; }
//
//  protected void initContext(Context ctx) {}
//
//  @Override
//  public void run(String... args) throws Exception {
//    Tomcat tomcat = new Tomcat();
//
//    tomcat.getServer().addLifecycleListener(new AprLifecycleListener());
//    tomcat.getServer().addLifecycleListener(new JreMemoryLeakPreventionListener());
//    tomcat.getServer().addLifecycleListener(new GlobalResourcesLifecycleListener());
//    tomcat.getServer().addLifecycleListener(new ThreadLocalLeakPreventionListener());
//
//    Connector connector = new Connector(getProtocol());
//    connector.setPort(getPort());
//    connector.setURIEncoding(Charsets.UTF_8.toString());
//    connector.setEnableLookups(false);
//
//    tomcat.getService().addConnector(connector);
//    tomcat.setConnector(connector);
//
//    Context context = tomcat.addWebapp(getContextPath(), new File(getResourceBase()).getAbsolutePath());
//    initContext(context);
//
//    log.info("Start Tomcat Web Server...");
//
//    tomcat.start();
//    tomcat.getServer().await();
//  }
//}

File: debop4k-web/src/main/kotlin/debop4k/web/tomcat/TomcatLauncher.kt
/*
 * Copyright (c) 2016. Sunghyouk Bae <sunghyouk.bae@gmail.com>
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package debop4k.web.tomcat

import debop4k.core.loggerOf
import org.apache.catalina.Context
import org.apache.catalina.connector.Connector
import org.apache.catalina.core.*
import org.apache.catalina.mbeans.GlobalResourcesLifecycleListener
import org.apache.catalina.startup.Tomcat
import org.springframework.boot.CommandLineRunner
import java.io.File

/**
 * TomcatLauncher
 * @author sunghyouk.bae@gmail.com
 */
open class TomcatLauncher : CommandLineRunner {

  private val log = loggerOf(javaClass)

  val port: Int get() = 8080
  val contextPath: String get() = "/"
  val resourceBase: String get() = "src/main/webapp"
  val protocol: String get() = "org.apache.coyote.http11.Http11Protocol"

  @Suppress("UNUSED_PARAMETER")
  protected fun initContext(ctx: Context) {
    // override method to additional initialization.
  }

  override fun run(vararg args: String?) {
    val tomcat = Tomcat().apply {
      server.addLifecycleListener(AprLifecycleListener())
      server.addLifecycleListener(JreMemoryLeakPreventionListener())
      server.addLifecycleListener(GlobalResourcesLifecycleListener())
      server.addLifecycleListener(ThreadLocalLeakPreventionListener())
    }

    val connector = Connector(protocol)
    connector.port = port
    connector.uriEncoding = Charsets.UTF_8.toString()
    connector.enableLookups = false

    tomcat.service.addConnector(connector)
    tomcat.connector = connector

    val context = tomcat.addWebapp(contextPath, File(resourceBase).absolutePath)
    initContext(context)

    log.info("Start Tomcat Web server... port={}, root={}", port, contextPath)

    tomcat.start()
    tomcat.server.await()
  }
}